cpp/src/arrow/io/buffered.cc

```diff
@@ -284,17 +285,30 @@ class BufferedInputStream::Impl : public BufferedBase {
 
   // Resize internal read buffer. Note that the internal buffer-size
   // should be not larger than the raw_read_bound_.
+  //
+  // SetBufferSize will not change the buffer_pos_ and bytes_buffered_.
   Status SetBufferSize(int64_t new_buffer_size) {
     if (new_buffer_size <= 0) {
       return Status::Invalid("Buffer size should be positive");
     }
     if ((buffer_pos_ + bytes_buffered_) >= new_buffer_size) {
-      return Status::Invalid("Cannot shrink read buffer if buffered data remains");
+      return Status::Invalid(
+          "Cannot shrink read buffer if buffered data remains, new_buffer_size:",
+          new_buffer_size, ", buffer_pos:", buffer_pos_,
+          ", bytes_buffered:", bytes_buffered_, ", buffer_size:", buffer_size_);
     }
     if (raw_read_bound_ >= 0) {
       // No need to reserve space for more than the total remaining number of bytes.
-      new_buffer_size = std::min(new_buffer_size,
-                                 bytes_buffered_ + (raw_read_bound_ - raw_read_total_));
+      if (bytes_buffered_ == 0) {
+        // Special case: we can not keep the current buffer because it does not
+        // contains any required data.
+        new_buffer_size = std::min(new_buffer_size, raw_read_bound_ - raw_read_total_);
+      } else {
+        // We should keeping the current buffer because it contains data that
+        // can be read.
+        new_buffer_size =
+            std::min(new_buffer_size, buffer_size_ + (raw_read_bound_ - raw_read_total_));
```

Comment:
```suggestion
        new_buffer_size =
            std::min(new_buffer_size, buffer_pos_ + buffer_size_ + (raw_read_bound_ - raw_read_total_));
```

Isn't `buffer_pos_` required to be kept as well?

Reply:
```
|     buffer_size_            |
|    buffer_pos_  |
```

Reply:
My bad! I have confused `bytes_buffered_` with `buffer_size_`.